### ISSUE TITLE
navbar app: eliminate loading lag

### ISSUE THREAD

**Issue by crisaless** [informatics-isi-edu/chaise#1936]
The app takes a while to load and is a bit of a distraction for users.

Thread from Slack:

karl: would it be simple or complicated to allow the top-level static content to get injected sooner, then have the slower dynamic stuff fill it in behind the scenes so a typical user would not notice much change?  i.e. I could imagine having the site name and top-level navbar menu names defined in the static chaise config file, and (re-)populate the actual nav menu content from the config fetched from the catalog annotation?

aref: when we added chaise-config support in catalog annotation, since we wanted to make sure we’re always getting the chaise-config before showing any UI, we changed the configuration of all the apps to wait for the catalog retrieval request and hence the noticable delay.
I think we could be able to refactor navbar and display it as soon as the app loads (using chaise-config.js) and then update it after catalog retrieval. But we need to make sure that the chaise-config properties like customCSS are already defined in chaise-config.js otherwise this would cause more confusing UI (for example the color of navbar might change before and after catalog retrieval). another option would be delaying the display of the rest of content so navbar latency wouldn’t be visible, but I don’t think that’s ideal. (edited) 

there are some other attributes that can cause problem. if we want to go with this, we should make sure we’re aware of the issues that it could cause. another types of chaise-config properties that comes to my mind are login/logout related ones.

**Comment by RFSH** [comment]
### Causes of the delay
The following are the extra requests that navbar is sending before showing its content:

- Fetch the JS and CSS files that the navbar relies on dynamically. The list of files is defined in `navbar.app.js`.
- When ermrest.js is fetched, it will send requests for all the dependencies of ERMrestJS and will dynamically fetch other JS files.
- Fetch the catalog and schema object by sending two requests to ERMrest.
- Fetch the custom CSS file dynamically.

### How to reduce the delay time

Based on the causes,

1. The first obvious change that we have to do is removing the unnecessary request to fetch the schema object. Only fetching the catalog should be enough for the configuration step. 

2. We can modify chaise and ermrestjs to only fetch the assets (JS and CSS files) that are not included in the HTML (and therefore prefetched). Then in the static sites, we can include these assets so they are prefetched. This will only improve the first time a user loads the page as these asset files will be cached by the browser.

To test how these will improve the performances, I tried them on facebase-dev. I copied the content of facebase-dev homepage and modified it a little bit. The following are the links to each one and how long it took to "load" the navbar.

|                                                                                                              | Chrome (ms) | Firefox (ms) | Safari (ms) | avg (ms) | avg imporv. (%) |
|--------------------------------------------------------------------------------------------------------------|-------------|--------------|-------------|----------|-----------------|
| [master](https://dev.facebase.org/~ashafaei/navbar/test/current.html)                                        | 1248.3      | 1345.5       | 1151        | 1248.2   |                 |
| [Only catalog](https://dev.facebase.org/~ashafaei/navbar/test/catalog.html)                                  | 742         | 828.6        | 585.9       | 718.8    | 42.41           |
| [Prefetch assets](https://dev.facebase.org/~ashafaei/navbar/test/prefetch.html)                              | 899.3       | 830          | 811.5       | 846.9    | 32.15           |
| [Only catalog and prefetch assets](https://dev.facebase.org/~ashafaei/navbar/test/prefetch-and-catalog.html) | 326.6       | 297.6        | 227.6       | 283.9    | 77.25           |

<details>
<summary>Click here for more info on the "load" time</summary>

- The reported time is not exactly the time that it takes to show the content. To make sure the time is comparable, I used the Javascript timer feature. I created a timer on the static site, and when the navbar.js file is executing the navbar list, I'm reporting the elapsed time.
- To make sure the time that I'm reporting is "fair", I tried this on Chrome, Firefox, and Safari (5 times for each one) and reported the average load time when I opened an incognito tab and loaded the URL.
- Obviously, the load times are affected by my internet speed and some load times might be more than the others just because my other devices at home were using the internet. So we should take these numbers with a grain of salt and the only take away from these results should be the pattern. 
</details>

Based on the time findings, we should at least remove the unnecessary request for the schema document. The prefetch improvement is helpful, but it might introduce new steps in the build process and it might not be worth it. I will need to investigate this further to find ways to make this simpler for automated builds. 

### Potential solutions

I can think of two ways of eliminating this visible lag and shift in the page:

1. Showing the navbar gradually.  In this approach, we could try and display the navbar as soon as we can, and then we should update it when the catalog request gets back. 
    - This requires changing how `chaise-config.js` is configured.
    - This solution will reduce the visible lag, but on its own, will not eliminate the visible shift. Because no matter what, to display the navbar, we have to at least fetch the navbar JS and CSS files dynamically (to eliminate the visible shift we need to prefetch the CSS and JS files in this approach). 
    - This solution also requires changes in how all the chaise (and deriva-web apps) are being configured and won't only affect the static sites.

2. Introduce the same state management that we have in chaise apps, in the static pages as well. This means that we can delay the display of the whole document and make sure we're displaying the navbar and the rest of the page together. 
    - This solution will eliminate the visible lag and shift of the page.
    - This will increase the page load time, so it will require reducing the navbar delay time as much as we can. 
    - This solution will be contained in static sites and won't require a lot of changes in the existing chaise setup. 


### Plan

Based on this these are the task that I'm working on: 

- [ ] Remove the unnecessary schema request during the configuration.
- [ ] Create a method that we can be used to signal the page that it can load (delay the whole page).
- [ ] Further, Investigate the prefetch of asset files.
- [ ] Investigate solution number 2 that is explained in the previous section.

**Comment by RFSH** [comment]
Regarding the JS and CSS static files that the navbar is currently loading, we could concatenate and minify those static files and reduce the HTTP load time this way.  We planned to do this as part of #1494, but we can start doing this for the assets that are required in the navbar app. Later we can do the same with the rest of the assets that are used in chaise apps.

**Comment by RFSH** [comment]
@crisaless 
I've completed the first and second steps of the plan that I've described in the issue. I reduced the delay and added an overlay loader to hide the content of the page and show the navbar and main content together. You can test it [using this link](https://dev.facebase.org/~ashafaei/navbar/test/new.html) (The loader that I'm showing can be easily customized and modified. I just wanted to show something instead of a blank white page).

The other steps might take a while and will have a lot of impact on deploy process, chaise, and deriva web apps; so it might not be feasible to implement them right away. I just wanted to get your feedback and if you agree I can create a PR with these changes. I'll explain the changes that you should apply to the navbar to use this new implementation when we merge the PR.
Later, we will implement the other steps of the plan to improve this one.

**Comment by RFSH** [comment]
The first two steps are implemented and merged with master. Please refer to [navbar app documentation](https://github.com/informatics-isi-edu/chaise/blob/master/lib/navbar/navbar-app.md#note) for more information on how you can delay the whole page until navbar is displayed.

**Comment by RFSH** [comment]
As part of #1935 and [ermrestjs#841](https://github.com/informatics-isi-edu/ermrestjs/pull/841), I did the two steps that I planned before which are

- introduce a better method of prefetching navbar dependencies. 
- minification of chaise dependencies.

I updated the [navbar app documentation]() to include more info and also how you can prefetch these navbar dependencies. Since we're using Jekyll in our deployments you can follow the instructions that I added in that document. 

To summarize, build scripts should be changed to copy the `dist/chaise-dependencies.html` file into the `_includes` folder of the static website. Then in the static pages, the `include` line to should be added to inject (prefetch) the dependencies.

@crisaless I noticed that the home page of dev.facebase is injecting ermrestjs. You don't need to manually inject it if dev.facebase build script is changed to take care of adding `dist/chaise-dependencies.html` file.

**Comment by RFSH** [comment]
I am going to close this issue. I created another issue for the remaining described task.
